Actually construct default vector values

lindsayad opened the pull request:
Previously default vector values were always `{0, 0, 0}`. Now we can actually supply default vector values in a manner similar to what is done for multiple coupled regular variables (see #11928)

lindsayad committed:
Actually construct default coupled vector values

lindsayad committed:
Make sure that _ad_default_vector_gradient is sized

lindsayad committed:
Test new default coupled vector value capability

Refs #10049

lindsayad committed:
Test for warning

lindsayad committed:
Test assigning default value from the source file

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/289819/) on 5d1bd3e wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12887/site)

This comment will be updated on new commits.

lindsayad committed:
Regold ad_vector_coupled_default

We are actually now using the default values that we are passing. The
previous gold file could actually be realized by commenting out the
convection block (e.g. it was doing nothing). We also need to
allow_warnings because we are not passing a LIBMESH_DIM sized vector

permcody commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -321,6 +322,27 @@ Coupleable::getDefaultVectorValue(const std::string & var_name)
+                ". Did you accidently leave something off? We are going to assign 0s, assuming "
+                "this "
+                "was intentional.");
+          already_warned = true;
```
Use mooseDoOnce macro instead - way easier.

lindsayad replied:
I chose not to use `mooseDoOnce` in case (in the admittedly unlikely event) I wanted to warn for multiple variables

permcody commented on framework/src/utils/InputParameters.C:
```diff
@@ -500,7 +500,7 @@ InputParameters::defaultCoupledValue(const std::string & coupling_name, unsigned
                "coupled value should have been properly guarded with isCoupled()\n");
 
-  return value_it->second._coupled_default[i];
+  return value_it->second._coupled_default.at(i);
```
Do we need this? I'm not a big fan of `at()` because it throws and in MPI land it doesn't call MPI_Abort which can complicate tracking failures down. We have the check above so do we need this?

lindsayad replied:
The above check only looks at whether there are *any* coupled values for the `coupling_name`; it doesn't look to see whether the component `i` is within the `_coupled_default` bounds. I could add another method like `hasDefaultCoupledComponent` that I check in the relevant `Coupleable` methods if you like that better

permcody commented on test/src/kernels/EFieldAdvection.C:
```diff
@@ -33,8 +33,9 @@ EFieldAdvection::EFieldAdvection(const InputParameters & parameters)
-    _vector_phi(_assembly.phi(_efield_var)),
+    _efield_coupled(isCoupled("efield")),
+    _efield_var(_efield_coupled ? getVectorVar("efield", 0) : nullptr),
+    _vector_phi(_efield_coupled ? &_assembly.phi(*_efield_var) : nullptr),
```
Another alternative is to reference a dummy object so you don't have to use a pointer in the member methods. Up to you.

lindsayad replied:
I kind of like the explicit nullness

permcody commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -321,6 +322,27 @@ Coupleable::getDefaultVectorValue(const std::string & var_name)
+      {
+        try
+        {
+          (*value)[qp](i) = _c_parameters.defaultCoupledValue(var_name, i);
```
Oh... so you did wrap the access of `at()`. I suppose this is ok, just not my favorite construct but valid.
